cpp/src/traversal/tsp_solver.hpp

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda.h>
+#include <curand_kernel.h>
+#include <limits.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <raft/cuda_utils.cuh>
+
+#include "tsp_utils.hpp"
+
+namespace cugraph {
+namespace detail {
+
+__device__ float *best_soln;
+__device__ int *best_route;
+extern __shared__ int shbuf[];
+
+__global__ void init(int *mylock, int *n_climbs, int *best_tour)
+{
+  *mylock    = 0;
+  *n_climbs  = 0;
+  *best_tour = INT_MAX;
+  best_soln  = nullptr;
+  best_route = nullptr;
+}
+
+__global__ __launch_bounds__(2048, 2) void two_opt_search(int *mylock,
```

Comment:
This is a quite large hand-written kernel. Two issues:
(1.) Couldn't be this designed using Thrust/CUB primitives?
(2.) Is there a possibility to split this kernel into smaller `__device__` functions, so that reading and maintaining this code later would become possible / easier ?